wgsl/index.bs

```diff
@@ -989,86 +988,365 @@ Host-shareable type |S| satisfies <dfn noexport>uniform buffer layout</dfn> when
-                   less than the number of components in *P*’s pointee type.
-                   The subaccess evaluation yields a pointer to the storage for
-                   the K’th component within P’s referenced storage,
-                   using zero-based indexing. If P's [=storage class=] is SC, and
-                   the K'th member of P's pointee type is of type T, then
-                   the result type is `ptr<SC,T>`.
-</table>
+### Forming a pointer value ### {#forming-a-pointer-value}
+
+When evaluating an expression, a pointer value is formed in one of the
+following ways:
+
+* By using the name of a variable that is in scope.
+    The result is the pointer to the storage allocated for the variable.
+* By a <dfn noexport>composite value pointer sub-access</dfn> expression:
+    * Given a pointer to a vector, appending a single-letter vector access phrase
+        results in a pointer to the named component of the vector.
+        See [[#component-pointer-from-vector-pointer]].
+    * Given a pointer to a vector, appending an array index access phrase
+        results in a pointer to the indexed component of the vector.
+        See [[#component-pointer-from-vector-pointer]].
+    * Given a pointer to a matrix, appending an array index access phrase
+        results in a pointer to the indexed column vector of the matrix.
+        See [[#matrix-access-expr]].
+    * Given a pointer to an array, appending an array index access phrase
+        results in a pointer to the indexed element of the array.
+        See [[#array-access-expr]].
+    * Given a pointer to a structure, appending a member access phrase
+        results in a pointer to the named member of the structure.
+        See [[#struct-access-expr]].
+* By passing a pointer value to a function, where the formal parameter of the function
+     has a matching pointer type.
+     The value denoted by the formal parameter inside the called function is the same
+     as the pointer value for the corresponding operand of the function call site
+     in the calling function.
+     See TODO.
+* By a `const` declaration with pointer type, where the initializer is a pointer
+     value formed in one of the above ways.
+     A later use of the declared name results in the pointer value that was computed
+     for the initializer when the declaration was executed.
+
+    <div class='example wgsl' heading='Pointer from a variable'>
+      <xmp highlight='rust'>
+        // Declare a variable in the private storage class, for storing an f32 value.
+        var<private> x: f32;
+
+        fn f() -> void {
+            // Declare a variable in the function storage class, for storing an i32 value.
+            var y: i32;
+
+            // An expression fragment using the name 'x' evaluates to a pointer of type
+            // ptr<private,f32>, pointing at the storage allocated for the variable 'x'.
+            const x_reference: ptr<private,f32> = x;
```

Comment:
this is a bit confusing, since we are no longer being able to reason about an expression type of "x". So it make the rules more complex, and goes against limited inference of types that we want to get eventually.
was it considered to use `&x` syntax instead?

Reply:
When you add address-of, you now have to distinguish between l-value and r-value.
When I look at a variable in isolation, I don't know if it's being used as an l-value or r-value until further out in nesting.   I have to defer evaluation (in my head) until I see or don't see a & operator used as address-of.  Also & is the same name for a binary operator.

Example
       &a.b.c.d[12]; 

In this example I would like to evaluate from what looks like the inside out (right to left) but I don't know if I'm dealing with pointers or values until I get to the very left.    There's an inherent ambiguity there just the same.

C and C++ also have an arrow operator to step down into a pointer-to-composite: `->` 

In the current design you stay in the pointer domain for as long as possible, i.e. dereference (for load) only when necessary.

Reply:
> When you add address-of, you now have to distinguish between l-value and r-value.

Right, and we have options there. I.e. we could even go with "store(my_pointer, value)", given that pointers are not widely used. This wouldn't affect variable assignments, obviously, since variables are not the pointers.

> In this example I would like to evaluate from what looks like the inside out (right to left) but I don't know if I'm dealing with pointers or values until I get to the very left. There's an inherent ambiguity there just the same.

Is this really a problem to parse? Both C and Rust have "&" for the address.
I haven't implemented it, but it seems like the "&" is clear based on the parser context. if an expression is expected, then "&" is an address. If other things are expected, then "&" is a binary operator.

Reply:
@litherum was that a reaction to "store(my_pointer, value)" (which I don't think is really necessary), or to having "&" in general for taking a pointer?